### DISCUSSION TITLE
Upload File to Git LFS using Rest API

### DISCUSSION THREAD

**Discussion by bauergeorg** [community/community#24623] [Apps, API and Webhooks]
<p>Hi, I using python PyGithub (REST API) to communicate with github and handle multi-repo stuff (releases and so on). Now we want to use Git LFS and I try to replace git LFS files.</p>
<p><strong>How I can upload a file to the github lfs storage (aws s3) using Rest API (in python)?</strong></p>
<p>Actual I can upload the file and change the git lfs pointer. But the link is crashed when I try to download the file.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/9/8/986a99d991f1bf656f423ca52d5582b4d8157797.png" data-download-href="/uploads/short-url/lKkXXkc4aZfYwnrwgcx0wmuolRd.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/79077565/181085705-b4dcb7a5-fb5b-4e6a-882f-3761268f0373.png" alt="image" data-base62-sha1="lKkXXkc4aZfYwnrwgcx0wmuolRd" width="690" height="393" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/9/8/986a99d991f1bf656f423ca52d5582b4d8157797_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">image</span><span class="informations">783×446 18.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p>(Hint: the files created by origin git lfs command I can download)</p>
<p>I tried a lots of python commands to upload the file. This is the only way it works without an checksum error.<br>
<code>requests.put(url=href, headers=header, data=data)</code></p>

**Comment by bauergeorg** [comment]
<p>To clone the total repository and using a git lfs command isn’t a option. The repository is too large.</p>

**Comment by bauergeorg** [comment]
<p>I try to explain what I do in detail:</p>
<ol>
<li>I have added some git lfs files with a git lfs client.</li>
<li>I try to download these files on github page. This works. Fine</li>
<li>Now i try to replace these files via Rest API<br>
3.1 get sha of branch in repo ($parent_sha)<br>
3.2 get old content of lfs pointer ($old_sha and $old_size)<br>
3.3 calculate content of new lfs pointer ($new_sha and $new_size)<br>
3.4 create new lfs pointer content (replace $old_sha with $new_sha and $old_size with $new_size)<br>
3.5 get url to upload file to lfs</li>
</ol>
<pre><code class="lang-auto">headers = {'Content-type': 'application/json', 'Accept': 'application/vnd.git-lfs+json'}
pointer_data = '{"operation": "upload", "transfer": ["basic"], "objects": [{"oid": "' + $new_sha + '", "size": ' + str($new_size) + '}]}'
url = 'https://github.com/{}/{}.git/info/lfs/objects/batch'.format(self.org, repo.name)
res = requests.post(url, headers=headers, data=pointer_data, auth=(self.token, ''))
ans1 = json.loads(res.text)
href = ans1['objects'][0]['actions']['upload']['href']
header = ans1['objects'][0]['actions']['upload']['header']
</code></pre>
<p>3.6 read new data (not encoded)<br>
<code>data = open(source_file_path, 'rb').read()</code><br>
3.7 send file to git lfs<br>
<code>ans2 = requests.put(url=href, headers=header, data=data)</code><br>
3.8 create blob with new content (lfs pointer)<br>
3.9 create a git tree to update the reference<br>
3.10 commit the tree<br>
3.11 update branch to point to new commit</p>
<p>I followed the instructions of: <a href="https://www.mattmoriarity.com/2019-04-25-uploading-media-with-git-lfs/#initiating-the-transfer" class="inline-onebox" rel="noopener nofollow ugc">Uploading Media with Git LFS | Matt Moriarity</a><br>
I think the problem is at step 3.7. the url that I get is a aws s3 server address. when I try to change my put command (to add an filename or something like that) I get a sha256-error. When I run my code example above I get no error (command code 200). But the link between github homepage and git lfs storage (aws s3) won’t work.</p>

**Comment by bauergeorg** [comment, marked as answer]
<p>I found a solution. I forgot to verify my upload.</p>
<p>python-based code extract using PyGithub:</p>
<pre><code class="lang-python"># see: https://www.mattmoriarity.com/2019-04-25-uploading-media-with-git-lfs/#initiating-the-transfer
  
# check for verify lock
headers = {'Content-Type': 'application/vnd.git-lfs+json', 'Accept': 'application/vnd.git-lfs+json'}
url = 'https://lfs.github.com/{}/{}/locks/verify'.format(self.org, repo.name)
  
# locks
ans0 = requests.post(url, headers=headers, auth=(self.token, ''))
if ans0.status_code != 200:
    print('status code: ' +  str(ans0.status_code))
    print('text: ' + ans0.text)
    raise Exception("Status code error")
  
# handle unhandled lock stuff
# we expect an no locked files (empty answer)
# see: https://github.com/git-lfs/git-lfs/blob/32d3ba7adc8a932ff5855964952ccb5cb4833a58/docs/api/locking.md
if ans0.text.find('{"ours":[],"theirs":[],"next_cursor":""}') == -1:
   raise Exception("Lock handle error. Please contact your favorite developer to add lfs lock handling!")
  
# print ans
#pprint(ans0.text)
  
# get old content
old_lfs_pointer_content = decoded_content.split('\n')
old_sha = old_lfs_pointer_content[1].replace('oid sha256:', '')
old_size = int(old_lfs_pointer_content[2].replace('size ', ''))
  
# calculate content of lfs pointer to replace
new_size = get_file_size(source_file_path)
new_sha = get_file_hash(source_file_path)
  
# replace content
new_lfs_pointer_content = decoded_content.replace(str(old_size), str(new_size))
new_lfs_pointer_content = new_lfs_pointer_content.replace(old_sha, new_sha)
  
# get url to upload file to lfs
headers = {'Content-Type': 'application/vnd.git-lfs+json', 'Accept': 'application/vnd.git-lfs+json'}
pointer_data = '{"operation": "upload", "transfer": ["basic"], "objects": [{"oid": "' + new_sha + '", "size": ' + str(new_size) + '}]}'
url = 'https://github.com/{}/{}.git/info/lfs/objects/batch'.format(self.org, repo.name)
res = requests.post(url, headers=headers, data=pointer_data, auth=(self.token, ''))
ans1 = json.loads(res.text)
pprint(ans1['objects'][0])
upload_href = ans1['objects'][0]['actions']['upload']['href']
upload_header = ans1['objects'][0]['actions']['upload']['header']
verify_href = ans1['objects'][0]['actions']['verify']['href']
verify_header = ans1['objects'][0]['actions']['verify']['header']
  
# add content type to header
upload_header['Content-Type'] = 'application/octet-stream'
# read new data (not encoded)
data = open(source_file_path, 'rb').read()

# send/upload file to git lfs
ans2 = requests.put(url=upload_href, headers=upload_header, data=data)
if ans2.status_code != 200:
   print('status code: ' +  str(ans2.status_code))
   print('text: ' + ans2.text)
   raise Exception("Status code error")
  
# print ans
pprint(ans2.text)
  
# verify
# see: https://github.com/git-lfs/git-lfs/blob/32d3ba7adc8a932ff5855964952ccb5cb4833a58/docs/api/basic-transfers.md#verification
verify_data = '{"oid": "' + new_sha + '", "size": ' + str(new_size) + '}'
ans3 = requests.post(url=verify_href, headers=verify_header, data=verify_data, auth=(self.token, ''))
if ans3.status_code != 200:
   print('status code: ' +  str(ans3.status_code))
   print('text: ' + ans3.text)
   raise Exception("Status code error")
  
# print ans
#pprint(ans3.text)

# create blob with new content (lfs pointer)
blob = repo.create_git_blob(content=new_lfs_pointer_content, encoding='utf-8')
</code></pre>

**Reply by sparrowt** [reply]
Many thanks! Github seems to have rendered that as markdown so here it is as a python codeblock for ease of reading/copying, along with some bonus comments:
```python
# check for verify lock

headers = {'Content-Type': 'application/vnd.git-lfs+json', 'Accept': 'application/vnd.git-lfs+json'}
url = 'https://lfs.github.com/{}/{}/locks/verify'.format(self.org, repo.name)

ans0 = requests.post(url, headers=headers, auth=(self.token, ''))
if ans0.status_code != 200:
    print('status code: ' +  str(ans0.status_code))
    print('text: ' + ans0.text)
    raise Exception("Status code error")

# see: https://github.com/git-lfs/git-lfs/blob/32d3ba7adc8a932ff5855964952ccb5cb4833a58/docs/api/locking.md
if ans0.text.find('{"ours":[],"theirs":[],"next_cursor":""}') == -1:
    raise Exception("Lock handle error. Please contact your favorite developer to add lfs lock handling!")

# get old content
# todo: fetch the content of an existing LFS file

# alternatively (or for a new file) just construct it yourself e.g. like this (new_size & new_sha are calculated below)
# lfs_pointer_content = (
#     "\n".join(
#         [
#             "version https://git-lfs.github.com/spec/v1",
#             f"oid sha256:{new_sha}",
#             f"size {new_size}",
#         ]
#     )
#     + "\n"
# )
old_lfs_pointer_content = decoded_content.split('\n')
old_sha = old_lfs_pointer_content[1].replace('oid sha256:', '')
old_size = int(old_lfs_pointer_content[2].replace('size ', ''))

# calculate content of lfs pointer to replace

new_size = get_file_size(source_file_path)
new_sha = get_file_hash(source_file_path)

# replace content

new_lfs_pointer_content = decoded_content.replace(str(old_size), str(new_size))
new_lfs_pointer_content = new_lfs_pointer_content.replace(old_sha, new_sha)

# get url to upload file to lfs
# https://github.com/git-lfs/git-lfs/blob/main/docs/api/batch.md

headers = {'Content-Type': 'application/vnd.git-lfs+json', 'Accept': 'application/vnd.git-lfs+json'}
pointer_data = '{"operation": "upload", "transfers": ["basic"], "objects": [{"oid": "' + new_sha + '", "size": ' + str(new_size) + '}]}'
url = 'https://github.com/{}/{}.git/info/lfs/objects/batch'.format(self.org, repo.name)
res = requests.post(url, headers=headers, data=pointer_data, auth=(self.token, ''))
ans1 = json.loads(res.text)
pprint(ans1['objects'][0])
upload_href = ans1['objects'][0]['actions']['upload']['href']
upload_header = ans1['objects'][0]['actions']['upload']['header']
verify_href = ans1['objects'][0]['actions']['verify']['href']
verify_header = ans1['objects'][0]['actions']['verify']['header']

# add content type to header

upload_header['Content-Type'] = 'application/octet-stream'

# read new data (not encoded)

data = open(source_file_path, 'rb').read()

# send/upload file to git lfs

ans2 = requests.put(url=upload_href, headers=upload_header, data=data)
if ans2.status_code != 200:
    print('status code: ' +  str(ans2.status_code))
    print('text: ' + ans2.text)
    raise Exception("Status code error")

# verify

# see: https://github.com/git-lfs/git-lfs/blob/32d3ba7adc8a932ff5855964952ccb5cb4833a58/docs/api/basic-transfers.md#verification
verify_data = '{"oid": "' + new_sha + '", "size": ' + str(new_size) + '}'
ans3 = requests.post(url=verify_href, headers=verify_header, data=verify_data, auth=(self.token, ''))
if ans3.status_code != 200:
    print('status code: ' +  str(ans3.status_code))
    print('text: ' + ans3.text)
    raise Exception("Status code error")

# create blob with new content (lfs pointer)
blob = repo.create_git_blob(content=new_lfs_pointer_content, encoding='utf-8')
# OR if you want to create and commit it all in one go then:
repo.create_file(
    path=target_path_within_repo,
    message=commit_message,
    content=new_lfs_pointer_content,
    branch=branch_name,
)
```

**Comment by AnesFoufa** [comment]
I wrote a simple library upload and commit a file to LFS using python. It is available on both Github and Pypi.
https://github.com/AnesFoufa/githublfs
https://pypi.org/project/githublfs/
